In [2]:
import pandas
import numpy as np
#подключение библиотек

Датасет 2salary.csv содержит различную информацию о зарплатах, такую как год выплаты зарплаты, уровень работы, тип работы, должность, зарплата за год, валюта, зарплата в USD, страна проживания, общий объем удаленной работы, страна главного офиса и среднее количество сотрудников.
Вывод первых 20 строк набора данных осуществляется с помощью функции head() библиотеки pandas.

In [3]:
df = pandas.read_csv('C:\\Users\\Alina\\Downloads\\2salary.csv',index_col=0)
#полученые данных из файла с датасетом в переменнную df
display(df.head(20))
#вывод первых 20 значений датасета

work_year experience_level employment_type                   job_title  \
0        2020               MI              FT              Data Scientist   
1        2020               SE              FT  Machine Learning Scientist   
2        2020               SE              FT           Big Data Engineer   
3        2020               MI              FT        Product Data Analyst   
4        2020               SE              FT   Machine Learning Engineer   
5        2020               EN              FT                Data Analyst   
6        2020               SE              FT         Lead Data Scientist   
7        2020               MI              FT              Data Scientist   
8        2020               MI              FT       Business Data Analyst   
9        2020               SE              FT          Lead Data Engineer   
10       2020               EN              FT              Data Scientist   
11       2020               MI              FT              Data Scientist   
12       2020               EN              FT              Data Scientist   
13       2020               MI              FT           Lead Data Analyst   
14       2020               MI              FT                Data Analyst   
15       2020               MI              FT                Data Analyst   
16       2020               EN              FT               Data Engineer   
17       2020               SE              FT           Big Data Engineer   
18       2020               EN              FT     Data Science Consultant   
19       2020               MI              FT          Lead Data Engineer   

        salary salary_currency  salary_in_usd employee_residence  \
0      70000.0             EUR          79833                 DE   
1     260000.0             USD         260000                 JP   
2      85000.0             GBP         109024                 GB   
3      20000.0             USD          20000                 HN   
4     150000.0             USD         150000                 US   
5      72000.0             USD          72000                 US   
6     190000.0             USD         190000                 US   
7   11000000.0             HUF          35735                 HU   
8     135000.0             USD         135000                 US   
9     125000.0             USD         125000                 NZ   
10     45000.0             EUR          51321                 FR   
11   3000000.0             INR          40481                 IN   
12     35000.0             EUR          39916                 FR   
13     87000.0             USD          87000                 US   
14     85000.0             USD          85000                 US   
15      8000.0             USD           8000                 PK   
16   4450000.0             JPY          41689                 JP   
17    100000.0             EUR         114047                 PL   
18    423000.0             INR           5707                 IN   
19     56000.0             USD          56000                 PT   

    remote_ratio company_location company_size  
0              0               DE            L  
1              0               JP            S  
2             50               GB            M  
3              0               HN            S  
4             50               US            L  
5            100               US            L  
6            100               US            S  
7             50               HU            L  
8            100               US            L  
9             50               NZ            S  
10             0               FR            S  
11             0               IN            L  
12             0               FR            M  
13           100               US            L  
14           100               US            L  
15            50               PK            L  
16           100               JP            S  
17           100               GB            S  
18   

Предметной областью анализа является набор данных с информацией о зарплате определенных IT-должностей с разным уровнем профессионализма и временем работы за 2020-2022 год в различных странах и компаниях. Файл содержит колонки: год выплаты зарплаты, уровень работы, тип работы, должность, зарплата за год, валюта, зарплата в USD, страна проживания, общий объем удаленной работы, страна главного офиса и среднее количество сотрудников.

In [4]:
df.info()
#оценка данных и пропущенных значений

<class 'pandas.core.frame.DataFrame'>
Int64Index: 610 entries, 0 to 606
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   work_year           610 non-null    int64  
 1   experience_level    610 non-null    object 
 2   employment_type     610 non-null    object 
 3   job_title           610 non-null    object 
 4   salary              607 non-null    float64
 5   salary_currency     609 non-null    object 
 6   salary_in_usd       610 non-null    int64  
 7   employee_residence  610 non-null    object 
 8   remote_ratio        610 non-null    int64  
 9   company_location    610 non-null    object 
 10  company_size        610 non-null    object 
dtypes: float64(1), int64(3), object(7)
memory usage: 57.2+ KB


По результатам использования метода info() для оценки данных можно сделать вывод, что столбец salary и salary_currency имеют пропущенные значения, а некоторые колонки имеют неправильный тип данных

In [5]:
df.columns
#названия столбцов для проверки на корректность

Index(['work_year', 'experience_level', 'employment_type', 'job_title',
       'salary', 'salary_currency', 'salary_in_usd', 'employee_residence',
       'remote_ratio', 'company_location', 'company_size'],
      dtype='object')

Данные столбцы не содержат повторов в названиях, а названия точно отображают содержимое данных столбцов, в них используется только один язык в змеином регистре, поэтому проблемы с названиями обнаружены не были.

Проверка пропущенных значений осуществляется с помощью метода isna(), а isna().sum() находит сумму пропущенных значений по столбцам.

In [6]:
print(df.isna().sum())
#сумма пропущенных значений по столбцам

work_year             0
experience_level      0
employment_type       0
job_title             0
salary                3
salary_currency       1
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
dtype: int64


Нужно удалить все строки со значениями NaN, используя метод dropna(), который удаляет все строки со значениями NaN.

In [7]:
df = df.dropna()
#удаление отсутствующих значений
print(df.isna().sum())
#проверка наличия отсутствующих значений

work_year             0
experience_level      0
employment_type       0
job_title             0
salary                0
salary_currency       0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
dtype: int64


Сумма пропущенных значений по всем столбцам равна нулю, что означает, что все пропущенные значения были удалены. 

Также необходимо проверить данные на наличие дубликатов, используя метод duplicated().sum() 

In [8]:
print(df.duplicated().sum())
#проверка наличия дубликатов

42


Так как дубликаты присутствуют, необходимо их удалить с помощью метода drop_duplicates().

In [9]:
df = df.drop_duplicates()
#удаление дубликатов
print(df.duplicated().sum())
#проверка наличия дубликатов после удаления

0


Удаление выполнено успешно, так как повторяющихся данных не осталось. После удаления строчек нужно обновить индексацию, чтобы в ней не осталось пропусков, используя метод reset_index(). Он создаст новый датафрейм, где индексы исходного датафрейма станут новой колонкой с названием index и все строки получат обычные индексы, но уже без пропусков.

In [10]:
df = df.drop_duplicates().reset_index()
#обновление индексации

Далее необходимо воспользоваться методом info() для проверки корректности типов данных по столбцам.

In [11]:
df.info()
#проверка типов данных

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 564 entries, 0 to 563
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   index               564 non-null    int64  
 1   work_year           564 non-null    int64  
 2   experience_level    564 non-null    object 
 3   employment_type     564 non-null    object 
 4   job_title           564 non-null    object 
 5   salary              564 non-null    float64
 6   salary_currency     564 non-null    object 
 7   salary_in_usd       564 non-null    int64  
 8   employee_residence  564 non-null    object 
 9   remote_ratio        564 non-null    int64  
 10  company_location    564 non-null    object 
 11  company_size        564 non-null    object 
dtypes: float64(1), int64(4), object(7)
memory usage: 53.0+ KB


По результатам проверки можно сделать вывод, что некоторые колонки имеют неправильный тип данных, который необходимо изменить для упрощения дальнейшей работы с данными. Для замены типа данных нужно воспользоваться методом astype().

In [12]:
df['experience_level'] = df['experience_level'].astype('string')
df['employment_type'] = df['employment_type'].astype('string')
df['job_title'] = df['job_title'].astype('string')
df['salary_currency'] = df['salary_currency'].astype('string')
df['employee_residence'] = df['employee_residence'].astype('string')
df['company_location'] = df['company_location'].astype('string')
df['company_size'] = df['company_size'].astype('string')
#изменеие типа данных «object» на «string»

Теперь необходимо проверить корректности типов данных по столбцам после их изменения.

In [13]:
df.info()
#проверка изменения

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 564 entries, 0 to 563
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   index               564 non-null    int64  
 1   work_year           564 non-null    int64  
 2   experience_level    564 non-null    string 
 3   employment_type     564 non-null    string 
 4   job_title           564 non-null    string 
 5   salary              564 non-null    float64
 6   salary_currency     564 non-null    string 
 7   salary_in_usd       564 non-null    int64  
 8   employee_residence  564 non-null    string 
 9   remote_ratio        564 non-null    int64  
 10  company_location    564 non-null    string 
 11  company_size        564 non-null    string 
dtypes: float64(1), int64(4), string(7)
memory usage: 53.0 KB


Столбец 'salary_in_usd' имеет строковый тип данных, который невозможно изменить из-за некорректного значения зарплаты в одной из строк. Для этого значение нужно изменить с помощью метода replace().

In [14]:
df['salary_in_usd'] = df['salary_in_usd'].replace('d210000', '210000')
#обработка некорректного значения данных

Теперь строковый тип данных можно изменить на числовой без ошибок.

In [15]:
df['salary_in_usd'] = df['salary_in_usd'].astype('int64')
#изменение типа «object» на «int»

Осуществить поиск неявных дубликатов можно с помощью метода unique().

In [16]:
for i in range(2, 12):
    a = df.columns[i]
    if df.dtypes[i] == 'string':
        print(df[a].unique())
#поиск неявных дубликатов по строковым значениям

<StringArray>
['MI', 'SE', 'EN', 'EX']
Length: 4, dtype: string
<StringArray>
['FT', 'CT', 'PT', 'FL']
Length: 4, dtype: string
<StringArray>
[                          'Data Scientist',
               'Machine Learning Scientist',
                        'Big Data Engineer',
                     'Product Data Analyst',
                'Machine Learning Engineer',
                             'Data Analyst',
                      'Lead Data Scientist',
                    'Business Data Analyst',
                       'Lead Data Engineer',
                        'Lead Data Analyst',
                            'Data Engineer',
                  'Data Science Consultant',
                          'BI Data Analyst',
                 'Director of Data Science',
                       'Research Scientist',
                 'Machine Learning Manager',
                 'Data Engineering Manager',
 'Machine Learning Infrastructure Engineer',
                              'ML Engineer',
   

Все некорректные названия нужно изменить, исправив опечатки тем же методом replace().

In [17]:
df['job_title'] = df['job_title'].replace('DataScientist', 'Data Scientist')
df['job_title'] = df['job_title'].replace('Data AnalyticsManager', 'Data Analytics Manager')
df['job_title'] = df['job_title'].replace('Data Analytics Lead', 'Lead Data Analyst')
df['job_title'] = df['job_title'].replace('Lead Data Analytics', 'Lead Data Analyst')
#обработка некорректного значения данных

Необходимо убедиться в отсутсвии неявных дубликатов.

In [18]:
for i in range(2, 12):
    a = df.columns[i]
    if df.dtypes[i] == 'string':
        print(df[a].unique())

<StringArray>
['MI', 'SE', 'EN', 'EX']
Length: 4, dtype: string
<StringArray>
['FT', 'CT', 'PT', 'FL']
Length: 4, dtype: string
<StringArray>
[                          'Data Scientist',
               'Machine Learning Scientist',
                        'Big Data Engineer',
                     'Product Data Analyst',
                'Machine Learning Engineer',
                             'Data Analyst',
                      'Lead Data Scientist',
                    'Business Data Analyst',
                       'Lead Data Engineer',
                        'Lead Data Analyst',
                            'Data Engineer',
                  'Data Science Consultant',
                          'BI Data Analyst',
                 'Director of Data Science',
                       'Research Scientist',
                 'Machine Learning Manager',
                 'Data Engineering Manager',
 'Machine Learning Infrastructure Engineer',
                              'ML Engineer',
   

Так как дубликаты найдены не были, можно приступать к созданию сводных таблиц. Необходимо создать сводную таблицу средней зарплаты по полному рабочему дню и частичной занятости с указанием страны компании.

In [20]:
table = pandas.pivot_table(df, index=['employment_type', 'company_location'], values=["salary_in_usd"], aggfunc=[np.mean])
display(table.query('employment_type == ["FL"] or employment_type == ["PT"]'))
#сводная таблица средней зарплаты полного рабочего дня и частичной занятости с указанием страны компании

mean
                                 salary_in_usd
employment_type company_location              
FL              PK                     12000.0
                US                     60000.0
PT              DE                     35461.5
                DK                     28609.0
                DZ                    100000.0
                ES                     10354.0
                IN                      5409.0
                IT                     21669.0
                NL                     69741.0
                US                     12000.0

Также необходимо сделать группировку средней зарплаты по уровню профессионализма с помощью groupby.

In [21]:
df_new = df.groupby('experience_level').agg(mean_sent=('salary_in_usd', 'mean')).reset_index()
display(df_new)
#группировка средней зарплаты по уровню опыта

experience_level      mean_sent
0               EN   61643.318182
1               EX  199392.038462
2               MI   87792.995192
3               SE  138544.095041

Сводная таблица средних зарплат по профессиям.

In [22]:
table1 = pandas.pivot_table(df, index=['job_title'], values=['salary_in_usd'], aggfunc=[np.mean])
display(table1)
#сводная таблица средней зарплаты по профессии

mean
                                          salary_in_usd
job_title                                              
3D Computer Vision Researcher               5409.000000
AI Scientist                               66135.571429
Analytics Engineer                        175000.000000
Applied Data Scientist                    175655.000000
Applied Machine Learning Scientist        142068.750000
BI Data Analyst                            74755.166667
Big Data Architect                         99703.000000
Big Data Engineer                          51974.000000
Business Data Analyst                      76691.200000
Cloud Data Engineer                       124647.000000
Computer Vision Engineer                   44419.333333
Computer Vision Software Engineer         105248.666667
Data Analyst                               89374.025000
Data Analytics Engineer                    64799.250000
Data Analytics Manager                    127134.285714
Data Architect                            177873.909091
Data Engineer                             109750.033058
Data Engineering Manager                  123227.200000
Data Science Consultant                    69420.714286
Data Science Engineer                      75803.333333
Data Science Manager                      158328.500000
Data Scientist                            103616.229008
Data Specialist                           165000.000000
Director of Data Engineering              156738.000000
Director of Data Science                  195074.000000
ETL Developer                              54957.000000
Finance Data Analyst                       61896.000000
Financial Data Analyst                    275000.000000
Head of Data                              160162.600000
Head of Data Science                      146718.750000
Head of Machine Learning                   79039.000000
Lead Data Analyst                         170402.250000
Lead Data Engineer                        139724.500000
Lead Data Scientist                       115190.000000
Lead Machine Learning Engineer             87932.000000
ML Engineer                               117504.000000
Machine Learning Developer                 85860.666667
Machine Learning Engineer                 101165.128205
Machine Learning Infrastructure Engineer  101145.000000
Machine Learning Manager                  117104.000000
Machine Learning Scientist                158412.500000
Marketing Data Analyst                     88654.000000
NLP Engineer                               37236.000000
Principal Data Analyst                    122500.000000
Principal Data Engineer                   328333.333333
Principal Data Scientist                  215242.428571
Product Data Analyst                       13036.000000
Research Scientist                        109019.500000
Staff Data Scientist                      105000.000000

По полученным сводным таблицам я могу сделать вывод, что самая высокая зарплата компаний за 2020-2022 в месяц за полный рабочий день в США, а по частичной занятости лидирует Алжир. Также самая высокая средняя зарплата за 2020-2022 в IT у Аналитика финансовых данных, она составляет 275000 долларов, а самая низкая средняя зарплата у исследователя компьютерного 3-d зрения, она составляет 5409 долларов. А также я проанализировала зависимость средней зарплаты от уровня профессионализма за 2020-2022 в IT и выяснила, что позиция джуниора наименее оплачиваемая и составляет 61643 долларов в месяц, а позиция эксперта наиболее оплачиваемая и составляет 199392 долларов в месяц.